### PR TITLE
Feature pulsar monitoring

### PR DESCRIPTION
**Pull request by liangyepianzhou**
image
image
image
image

* [ ]  If this is non-trivial feature, paste the links/URLs to the design doc.
* [ ]  Update the documentation to include this new feature.
* [ ]  Tests(including UT, IT, E2E) are added to verify the new feature.
* []  If it's UI related, attach the screenshots below.
* [ ]  If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #.
* [x]  Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

File: .github/workflows/skywalking.yaml [hunk 1]
```diff
@@ -641,6 +641,9 @@ jobs:
             config: test/e2e-v2/cases/kafka/kafka-monitoring/e2e.yaml
           - name: MQE Service
             config: test/e2e-v2/cases/mqe/e2e.yaml
+          - name: Pulsar
+            // TODO: test/e2e-v2/cases/pulsar/e2e.yaml
```
Comment (wu-sheng) [inline .github/workflows/skywalking.yaml:645]: Please the clean the codes and remove the TODO.

**Comment by wu-sheng** [comment]
Please ref https://github.com/apache/skywalking-booster-ui/pull/318 to submit PR for the translations of the menu words.

**Comment by wu-sheng** [comment]
Once this is done, please attach the screenshots. One more optional step is adding the monitoring in the showcase repository.

**Comment by wu-sheng** [comment]
CI Licence header check fails,

```
ERROR the following files don't have a valid license header: 
oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-bookkeeper.yaml
oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml
oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-cluster.yaml
test/e2e-v2/cases/pulsar/pulsar-cases.yaml 
ERROR one or more files does not have a valid license header 
```

And e2e all fails, please make sure you have run the tests locally through e2e tools, and submit the update.

File: test/e2e-v2/cases/pulsar/docker-compose.yml [hunk 2]
```diff
@@ -122,7 +130,7 @@ services:
     command: bash -c "bin/apply-config-from-env.py conf/broker.conf && exec bin/pulsar broker"
 
   otel-collector:
-    image: otel/opentelemetry-collector:0.72.0
+    image: otel/opentelemetry-collector:0.50.0
```
Comment (wu-sheng) [inline test/e2e-v2/cases/pulsar/docker-compose.yml:133]: Why downgrade the version? Is something wrong for 0.72?

File: test/e2e-v2/cases/pulsar/docker-compose.yml [hunk 3]
```diff
@@ -22,8 +22,16 @@ services:
       service: oap
     ports:
       - "12800:12800"
+    environment:
+        SW_ENABLE_UPDATE_UI_TEMPLATE: true
+        SW_STORAGE: mysql
+        SW_JDBC_URL: jdbc:mysql://mysql:3306/swtest?allowMultiQueries=true
```
Comment (wu-sheng) [inline test/e2e-v2/cases/pulsar/docker-compose.yml:28]: May I ask, why suddenly test in the MySQL storage?

Comment (liangyepianzhou) [reply test/e2e-v2/cases/pulsar/docker-compose.yml:28]: It is used to debug. I push it to GitHub to share with @liuhaoyang.

File: test/e2e-v2/cases/pulsar/docker-compose.yml (hunk 2 above)
Comment (liangyepianzhou) [reply test/e2e-v2/cases/pulsar/docker-compose.yml:133]: It does not match the config in `otel-collector-config.yaml`.
```
  logging:
    logLevel: debug
```

File: oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml [hunk 4]
```diff
@@ -34,66 +34,14 @@ metricPrefix: meter_pulsar_broker
...
-    exp: pulsar_ml_cache_entries.sum(['cluster', 'broker'])
-  - name: cache_hits_rate
-    exp: pulsar_ml_cache_hits_rate.sum(['cluster', 'broker'])
-  - name: cache_hits_throughput
-    exp: pulsar_ml_cache_hits_throughput.sum(['cluster', 'broker'])
-  - name: cache_misses_rate
-    exp: pulsar_ml_cache_misses_rate.sum(['cluster', 'broker'])
-  - name: cache_misses_throughput
-    exp: pulsar_ml_cache_misses_throughput.sum(['cluster', 'broker'])
-
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml]: Why are these removed?

Comment (liangyepianzhou) [reply oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml]: Pulsar only has 10 metrics with broker label.
image

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml]: Then how these deleted metrics look like?

Comment (liangyepianzhou) [reply oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml]: They only have the cluster label.

Comment (wu-sheng) [reply oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml]: I have discussed this with your mentor. Could you sync with him? OTEL collector could add the source as an extra label.
Bookeeper should be lands a new layer and in the database catalog.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/Layer.java [hunk 5]
```diff
@@ -193,7 +193,16 @@ public enum Layer {
...
+    KAFKA(31, true),
+
+    /**
+     * Pulsar is a distributed pub-sub messaging platform that provides high-performance, durable messaging.
+     * It is used to publish and subscribe to streams of records.
+     * Pulsar supports scalable and fault-tolerant messaging, making it suitable for use in distributed systems.
+     */
+    PULSAR(32, true),
+
+    BOOKKEEPER(33, true);
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/Layer.java:205]: Please add comments when you have time.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/analysis/Layer.java:205]: And documents should be updated as well.

**Comment by wu-sheng** [comment]
image

Recommend using `MB` at least to show the size. I think, from today's infra resource, less than 1MB is not worth showing as an increment.

File: docs/menu.yml [hunk 6]
```diff
@@ -265,6 +265,9 @@ catalog:
                 path: "/en/setup/backend/backend-rabbitmq-monitoring"
               - name: "Kafka"
                 path: "/en/setup/backend/backend-kafka-monitoring"
+                // TODO: backend-pulsar-monitoring
+              - name: "Pulsar"
+                path: "/en/setup/backend/backend-pulsar-monitoring"
           - name: "Self Observability"
```
Comment (liuhaoyang) [inline docs/menu.yml:271]: Did you forget to add the bookkeeper path?

File: oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml (hunk 4 above)
Comment (liuhaoyang) [reply oap-server/server-starter/src/main/resources/otel-rules/pulsar/pulsar-broker.yaml]: > I have discussed this with your mentor. Could you sync with him? OTEL collector could add the source as an extra label. Bookeeper should be lands a new layer and in the database catalog.

already done.

File: docs/menu.yml [hunk 7]
```diff
@@ -265,6 +265,10 @@ catalog:
                 path: "/en/setup/backend/backend-rabbitmq-monitoring"
               - name: "Kafka"
                 path: "/en/setup/backend/backend-kafka-monitoring"
+              - name: "Pulsar"
+                path: "/en/setup/backend/backend-pulsar-monitoring"
+              - name: "Bookkeeper"
+                path: "/en/setup/backend/backend-bookkeeper-monitoring"
```
Comment (wu-sheng) [inline docs/menu.yml:271]: `Bookkeeper` -> `BookKeeper`. Case is sensitive for the name.

According to its website, 
> About
Apache BookKeeper - a scalable, fault tolerant and low latency storage service optimized for append-only workloads

This part should be in the database catalog.

File: oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/management/ui/template/UITemplateInitializer.java [hunk 8]
```diff
@@ -42,43 +44,45 @@
...
     };
     private final UITemplateManagementService uiTemplateManagementService;
     private final ObjectMapper mapper;
 
     public UITemplateInitializer(ModuleManager manager) {
         this.uiTemplateManagementService = manager.find(CoreModule.NAME)
-                                                  .provider()
-                                                  .getService(UITemplateManagementService.class);
+                .provider()
+                .getService(UITemplateManagementService.class);
```
Comment (wu-sheng) [inline oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/management/ui/template/UITemplateInitializer.java:85]: Please don't reformat irrelative codes.

Comment (wu-sheng) [reply oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/management/ui/template/UITemplateInitializer.java:85]: Including all others in this file.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/bookkeeper/bookkeeper-cluster.json [hunk 9]
```diff
@@ -0,0 +1,393 @@
+[
+  {
+    "id": "Bookkeeper-Cluster",
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/bookkeeper/bookkeeper-cluster.json:3]: `Bookkeeper` -> `BookKeeper`

Please keep the name correctly in all things of your PR.

File: .github/workflows/skywalking.yaml [hunk 10]
```diff
@@ -641,6 +641,8 @@ jobs:
             config: test/e2e-v2/cases/kafka/kafka-monitoring/e2e.yaml
           - name: MQE Service
             config: test/e2e-v2/cases/mqe/e2e.yaml
+          - name: Pulsar
+            config: test/e2e-v2/cases/pulsar/e2e.yaml
```
Comment (wu-sheng) [inline .github/workflows/skywalking.yaml:645]:
```suggestion
          - name: Pulsar and BookKeeper
            config: test/e2e-v2/cases/pulsar/e2e.yaml
```

We are expecting you have tests/validation for metrics of Pulsar and BookKeeper.

**Comment by wu-sheng** [comment]
Notice, that the OSPP is reaching the deadline. SkyWalking community would not accept the PR until it is qualified. 
Only two days left, and I can't be sure your PR can be merged at the last minute.

**Comment by liangyepianzhou** [comment]
> Notice, that the OSPP is reaching the deadline. SkyWalking community would not accept the PR until it is qualified. Only two days left, and I can't be sure your PR can be merged at the last minute.

@wu-sheng This PR is basically ready. Please help review it.

File: docs/menu.yml (hunk 7 above)
Comment (wu-sheng) [reply docs/menu.yml:271]: @liangyepianzhou This comment has not been resolved.

File: docs/en/setup/backend/backend-pulsar-monitoring.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
+# Pulsar monitoring
+
+SkyWalking leverages Prometheus to collect metrics data from the Pulsar and leverages OpenTelemetry Collector to transfer the metrics to
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-pulsar-monitoring.md:3]:
```suggestion
SkyWalking leverages OpenTelemetry Collector to collect metrics data in Prometheus format from the Pulsar and transfer the metrics to
```

Please update two docs about this. We don't leverage Prometheus. We only use the Prometheus endpoint of Pulsar and BookKeeper only.

**Comment by wu-sheng** [comment]
@wankai123 Could you another round of review in case of missing anything?

File: test/e2e-v2/cases/pulsar/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+      file: ../../script/docker-compose/base-compose.yml
+      service: oap
+    ports:
+      - "12800:12800"
+    networks:
+      - e2e
+
+  # Start zookeeper
+  zookeeper:
+    image: zookeeper:latest
```
Comment (wankai123) [inline test/e2e-v2/cases/pulsar/docker-compose.yml:30]: suggest to use a specific version

File: test/e2e-v2/cases/pulsar/otel-collector-config.yaml [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+exporters:
+  otlp:
+    # The OAP Server address
+    endpoint: "oap:11800"
+    # The config format of OTEL version prior to 0.34.0, eg. 0.29.0, should be:
+    # insecure: true
+    tls:
+      insecure: true
+    #insecure: true
```
Comment (wankai123) [inline test/e2e-v2/cases/pulsar/otel-collector-config.yaml:53]:
```suggestion
```

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+                    "showSymbol": true,
+                    "showXAxis": true,
+                    "showYAxis": true,
+                    "legend": {
+                      "asTable": false
+                    }
+                  },
+                  "widget": {
+                    "name": "storage_size",
+                    "title": "Storage Size"
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:405]: please add the unit of the storage size

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+                    "type": "Line",
+                    "step": false,
+                    "smooth": false,
+                    "showSymbol": true,
+                    "showXAxis": true,
+                    "showYAxis": true
+                  },
+                  "widget": {
+                    "name": "storage_logical_size",
+                    "title": "Storage Logical Size"
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:340]: please add the unit of the storage size

File: docs/en/setup/backend/backend-bookkeeper-monitoring.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+| Bookie Ledgers Count         | meter_bookkeeper_bookie_ledgers_count                            | The number of the bookie ledgers.                          | Bookkeeper Cluster  |
+| Bookie Ledger Writable Dirs  | meter_bookkeeper_bookie_ledger_writable_dirs                     | The number of writable directories in the bookie.          | Bookkeeper Cluster  |
+| Bookie Ledger Dir Usage      | meter_bookkeeper_bookie_ledger_dir_data_bookkeeper_ledgers_usage | The number of successfully created connections.            | Bookkeeper Cluster  |
+| Bookie Entries Count         | meter_bookkeeper_bookie_entries_count                            | The number of failed connections.                          | Bookkeeper Cluster  |
+| Bookie Write Cache Size      | meter_bookkeeper_bookie_write_cache_size                         | The total number of closed connections.                    | Bookkeeper Cluster  |
+| bookie_write_cache_count     | meter_bookkeeper_bookie_write_cache_count                        | The usage of jvm buffer pool.                              | Bookkeeper Cluster  |
```
Comment (wankai123) [inline docs/en/setup/backend/backend-bookkeeper-monitoring.md:36]:
```suggestion
| Bookie Write Cache Count    | meter_bookkeeper_bookie_write_cache_count                        | The usage of jvm buffer pool.                              | Bookkeeper Cluster  |
```

Comment (wankai123) [reply docs/en/setup/backend/backend-bookkeeper-monitoring.md:36]: I'm a little confused `Bookie Write Cache Size` means `The total number of closed connections. `

File: docs/en/setup/backend/backend-bookkeeper-monitoring.md [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+| Bookie Read Cache Size       | meter_bookkeeper_bookie_read_cache_size                          | Time spent in a given JVM garbage collector in seconds.    | Bookkeeper Cluster  |
+| Bookie Write Cache Count     | meter_bookkeeper_bookie_read_cache_count                         | Time spent in a given JVM garbage collector in 30 seconds. | Bookkeeper Cluster  |
```
Comment (wankai123) [inline docs/en/setup/backend/backend-bookkeeper-monitoring.md:38]: It should be `Bookie Read Cache Count` ?

Comment (wu-sheng) [reply docs/en/setup/backend/backend-bookkeeper-monitoring.md:38]: @liangyepianzhou Notice, this is the processed metric. So, let's give a clear explanation. Count/counter is how to measure, it is not a proper name.

And when you fix these, you should fix the UI settings accordingly.

File: docs/en/setup/backend/backend-bookkeeper-monitoring.md [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+| Bookie Read Rate             | meter_bookkeeper_bookie_read_rate                                | Time spent in a given JVM garbage collector in 30 seconds. | Bookkeeper Cluster  |
+| Bookie Write Rate            | meter_bookkeeper_bookie_write_rate                               | Time spent in a given JVM garbage collector in 30 seconds. | Bookkeeper Cluster  |
```
[the remark this answers is not in the document]
Comment (wankai123) [reply docs/en/setup/backend/backend-bookkeeper-monitoring.md:40]: The descriptions are the same, please fix

File: oap-server/server-starter/src/main/resources/otel-rules/bookkeeper/bookkeeper-cluster.yaml [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+    exp: bookie_read_cache_size.sum(['cluster', 'node'])
+  - name: bookie_read_cache_count
+    exp: bookie_read_cache_count.sum(['cluster', 'node'])
+
+  - name: bookie_write_cache_hits
+    exp: bookie_write_cache_hits.sum(['cluster', 'node']).rate('PT1M')
+  - name: bookie_read_cache_hits
+    exp: bookie_read_cache_hits.sum(['cluster', 'node']).rate('PT1M')
+  - name: bookie_write_cache_misses
+    exp: bookie_write_cache_hits.sum(['cluster', 'node']).rate('PT1M')
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/otel-rules/bookkeeper/bookkeeper-cluster.yaml:64]: should be `bookie_write_cache_misses.sum(['cluster', 'node']).rate('PT1M')`

File: oap-server/server-starter/src/main/resources/otel-rules/bookkeeper/bookkeeper-node.yaml [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+    exp: jvm_buffer_pool_used_bytes.sum(['cluster', 'node', 'pool'])
+  - name: jvm_buffer_pool_capacity_bytes
+    exp: jvm_buffer_pool_capacity_bytes.sum(['cluster', 'node', 'pool'])
+  - name: jvm_buffer_pool_used_buffers
+    exp: jvm_buffer_pool_used_buffers.sum(['cluster', 'node', 'pool'])
+
+  - name: jvm_gc_collection_seconds_count
+    exp: jvm_gc_collection_seconds_count.sum(['cluster', 'node', 'gc']).rate('PT1M')
+  - name: jvm_gc_collection_seconds_sum
+    exp: jvm_gc_collection_seconds_sum.sum(['cluster', 'node', 'gc']).rate('PT1M')
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/otel-rules/bookkeeper/bookkeeper-node.yaml:81]: I think `rate` is not a good calculation for `gc seconds count/sum`. 
please check `rate` definition: https://skywalking.apache.org/docs/main/next/en/concepts-and-designs/mal/#rate

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json (hunk 14 above)
Comment (liangyepianzhou) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:405]: Should we add the metrics in the title?

File: oap-server/server-starter/src/main/resources/otel-rules/bookkeeper/bookkeeper-cluster.yaml (hunk 19 above)
Comment (liangyepianzhou) [reply oap-server/server-starter/src/main/resources/otel-rules/bookkeeper/bookkeeper-cluster.yaml:64]: This should be deleted.

**Comment by liangyepianzhou** [comment]
@wu-sheng @liuhaoyang I think we should give a warning to latecomers: 
1.  The current tool `compose` in `[skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e)` does not support volumes.
2. The container that is depended on cannot have a command.

This problem is inconsistent in the operation of local and CI. This took me a lot of time to test and verify.

**Comment by wu-sheng** [comment]
> @wu-sheng @liuhaoyang I think we should give a warning to latecomers: 
> 
> 1.  The current tool `compose` in `[skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e)` does not support volumes.
> 
> 2. The container that is depended on cannot have a command.
> 
> 
> 
> This problem is inconsistent in the operation of local and CI. This took me a lot of time to test and verify.

Forwars to @kezhenxu94 @fgksgf Any comment about this?

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json [hunk 21]
```diff
@@ -0,0 +1,447 @@
...
+                  "id": "16-0-18",
+                  "metricTypes": [
+                    "readLabeledMetricsValues"
+                  ],
+                  "metrics": [
+                    "meter_pulsar_storage_size/1024/1024"
+                  ],
+                  "metricMode": "General",
+                  "moved": false,
+                  "expressions": [],
```
Comment (liuhaoyang) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:390]: image
`expression` is empty, causing the `storage size` graph to load exceptionally.

**Comment by kezhenxu94** [comment]
> @wu-sheng @liuhaoyang I think we should give a warning to latecomers: 
> 
> 1.  The current tool `compose` in `[skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e)` does not support volumes.
> 
> 2. The container that is depended on cannot have a command.
> 
> 
> 
> This problem is inconsistent in the operation of local and CI. This took me a lot of time to test and verify.

Can you provide some examples (docker-compose.yaml) for the cases above?

**Comment by liangyepianzhou** [comment]
> > @wu-sheng @liuhaoyang I think we should give a warning to latecomers:
> > 
> > 1. The current tool `compose` in `[skywalking-infra-e2e](https://github.com/apache/skywalking-infra-e2e)` does not support volumes.
> > 2. The container that is depended on cannot have a command.
> > 
> > This problem is inconsistent in the operation of local and CI. This took me a lot of time to test and verify.
> 
> Can you provide some examples (docker-compose.yaml) for the cases above?

For volume: https://github.com/apache/skywalking/pull/11339/commits/90703434d32dc0dbf51d11dfada98b4ecdc1f74c
- The test passed locally, but the cI report `ERROR /bin/sh command not executable`
For depend on: https://github.com/apache/skywalking/pull/11339/commits/1b41304c5499836963098b58dcd27880b883fc02#diff-cd52bb11901c8fa7de29792681bee6848ce71ae7a1c0e9a3f4b4058f07e84303L16
- When I tested locally, the cluster status was healthy, but CI reported an error `ERROR: for bookie  Container "bfe1184eba80" is unhealthy.ERROR: for pulsar-init  Container "bfe1184eba80" is unhealthy.`

**Comment by wu-sheng** [comment]
@liangyepianzhou Besides discussion around e2e, are all other comments resolved?

File: docs/menu.yml (hunk 7 above)
Comment (liangyepianzhou) [reply docs/menu.yml]: Fixed

File: docs/en/setup/backend/backend-bookkeeper-monitoring.md (hunk 17 above)
Comment (liangyepianzhou) [reply docs/en/setup/backend/backend-bookkeeper-monitoring.md]: Fixed

**Comment by liangyepianzhou** [comment]
> @liangyepianzhou Besides discussion around e2e, are all other comments resolved?

There is still one [comment](https://github.com/apache/skywalking/pull/11339#discussion_r137545068); I am still determining where the unit should be added. Title or other special place.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json (hunk 14 above)
Comment (wankai123) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:407]: No, I mean need a `Unit` to indicate `MB/GB`

**Comment by wu-sheng** [comment]
> > @liangyepianzhou Besides discussion around e2e, are all other comments resolved?
> 
> There is still one [comment](https://github.com/apache/skywalking/pull/11339#discussion_r137545068); I am still determining where the unit should be added. Title or other special place.

Your link doesn't work. Which one are you referring to?

**Comment by wankai123** [comment]
> > @liangyepianzhou Besides discussion around e2e, are all other comments resolved?
> 
> There is still one [comment](https://github.com/apache/skywalking/pull/11339#discussion_r137545068); I am still determining where the unit should be added. Title or other special place.

you can refer to the other dashboard templates, I think add on the title is good.

**Comment by liangyepianzhou** [comment]
@wu-sheng @wankai123 @liuhaoyang All the comments are addressed. Could you please take a look? We are asked to merge this PR tonight.

File: docs/en/changes/changes.md [hunk 22]
```diff
@@ -13,6 +13,7 @@
 * ElasticSearchClient: Add `deleteById` API.
 * Fix Custom alarm rules are overwritten by 'resource/alarm-settings.yml'
 * Support Kafka Monitoring.
+* Support Pulsar Monitoring.
```
Comment (wu-sheng) [inline docs/en/changes/changes.md:16]:
```suggestion
* Support Pulsar server and BookKeeper server Monitoring.
```

File: docs/en/setup/backend/backend-bookkeeper-monitoring.md [hunk 23]
```diff
@@ -0,0 +1,61 @@
+# BookKeeper monitoring
+
+SkyWalking leverages OpenTelemetry Collector to collect metrics data from the BookKeeper and leverages OpenTelemetry Collector to transfer the metrics to
+[OpenTelemetry receiver](opentelemetry-receiver.md) and into the [Meter System](./../../concepts-and-designs/meter.md).
+Kafka entity as a `Service` in OAP and on the `Layer: BOOKKEEPER.
+
+## Data flow
+
+1. BookKeeper exposes metrics in Prometheus format.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-bookkeeper-monitoring.md:9]:
```suggestion
1. BookKeeper exposes metrics through Prometheus endpoint.
```

File: docs/en/setup/backend/backend-pulsar-monitoring.md [hunk 24]
```diff
@@ -0,0 +1,67 @@
+# Pulsar monitoring
+
+SkyWalking leverages OpenTelemetry Collector to collect metrics data in Prometheus format from the Pulsar and transfer the metrics to
+[OpenTelemetry receiver](opentelemetry-receiver.md) and into the [Meter System](./../../concepts-and-designs/meter.md).
+Kafka entity as a `Service` in OAP and on the `Layer: PULSAR.
+
+## Data flow
+
+1. Pulsar exposes metrics in Prometheus format.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-pulsar-monitoring.md:9]:
```suggestion
1. Pulsar exposes metrics through Prometheus endpoint.
```

File: docs/menu.yml [hunk 25]
```diff
@@ -259,12 +259,16 @@ catalog:
                 path: "/en/setup/backend/backend-elasticsearch-monitoring"
               - name: "MongoDB"
                 path: "/en/setup/backend/backend-mongodb-monitoring"
+              - name: "Bookkeeper"
```
Comment (wu-sheng) [inline docs/menu.yml:262]:
```suggestion
              - name: "BookKeeper"
```

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/bookkeeper/bookkeeper-root.json [hunk 26]
```diff
@@ -0,0 +1,37 @@
+[
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/bookkeeper/bookkeeper-root.json:1]: This root dashboard doesn't have a widget pointing to the doc of latest. Please refer to the other existing dashboards.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-root.json [hunk 27]
```diff
@@ -0,0 +1,37 @@
+[
```
Comment (wu-sheng) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-root.json:1]: This root dashboard doesn't have a widget pointing to the doc of latest. Please refer to the other existing dashboards.

Comment (liangyepianzhou) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-root.json:1]: Fixed.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/bookkeeper/bookkeeper-root.json (hunk 26 above)
Comment (liangyepianzhou) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/bookkeeper/bookkeeper-root.json:1]: Fixed.

File: docs/menu.yml (hunk 25 above)
Comment (liangyepianzhou) [reply docs/menu.yml]: Fixed.

**Comment by wankai123** [comment]
In the root dashboards, the description text should be on the top, please check and fix both of 2.
image

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json [hunk 28]
```diff
@@ -0,0 +1,447 @@
...
+                    "showSymbol": true,
+                    "showXAxis": true,
+                    "showYAxis": true
+                  },
+                  "widget": {
+                    "title": "Message Rate In",
+                    "tips": "Incoming message rate.",
+                    "name": "message_rate_in"
+                  },
+                  "expressions": [],
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:169]: Please use MQE and check all dashboards config.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json (hunk 15 above)
Comment (wankai123) [reply oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-cluster.json:340]: please don't mark resolved unless you really fix it

File: docs/en/setup/backend/backend-pulsar-monitoring.md [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+| Throughput In        | meter_pulsar_throughput_in                 | The total throughput coming into this cluster (byte per second).                                       | Pulsar Cluster |
+| Throughput Out       | meter_pulsar_throughput_out                | The total throughput going out from this cluster (byte per second).                                    | Pulsar Cluster |
+| Storage Size         | meter_pulsar_storage_size                  | The total storage size of all topics in this broker (in bytes).                                        | Pulsar Cluster |
+| Storage Logical Size | meter_pulsar_storage_logical_size          | The storage size of all topics in this broker without replicas (in bytes).                             | Pulsar Cluster |
+| Storage Write Rate   | meter_pulsar_storage_write_rate            | The total message batches (entries) written to the storage for this broker (message batch per second). | Pulsar Cluster |
+| Storage Read Rate    | meter_pulsar_storage_read_rate             | The total message batches (entries) read from the storage for this broker (message batch per second).  | Pulsar Cluster |
```
Comment (wankai123) [inline docs/en/setup/backend/backend-pulsar-monitoring.md:42]: The units such as `(message per second)`  you add to the description should be added to the dashboards too.

File: oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-broker.json [hunk 30]
```diff
@@ -0,0 +1,422 @@
...
+            "type": "Line",
+            "step": false,
+            "smooth": false,
+            "showSymbol": true,
+            "showXAxis": true,
+            "showYAxis": true
+          },
+          "widget": {
+            "name": "jvm_gc_collection_seconds_sum",
+            "title": "GC Time (ms/min)"
```
Comment (wankai123) [inline oap-server/server-starter/src/main/resources/ui-initialized-templates/pulsar/pulsar-broker.json:381]: Check the unit whether it is correct

File: docs/en/setup/backend/backend-bookkeeper-monitoring.md [hunk 31]
```diff
@@ -32,12 +32,12 @@ the OAP. In each cluster, the nodes are represented as `Instance`.
...
+| Bookie Read Cache Size         | meter_bookkeeper_bookie_read_cache_size                          | The size of the bookie read cache (MB).           | Bookkeeper Cluster  |
 | Bookie Read Cache Entry Count  | meter_bookkeeper_bookie_read_cache_count                         | The entry count in the bookie read cache.         | Bookkeeper Cluster  |
-| Bookie Read Rate               | meter_bookkeeper_bookie_read_rate                                | The bookie read rate.                             | Bookkeeper Cluster  |
-| Bookie Write Rate              | meter_bookkeeper_bookie_write_rate                               | The bookie write rate.                            | Bookkeeper Cluster  |
+| Bookie Read Rate               | meter_bookkeeper_bookie_read_rate                                | The bookie read rate (MB/min).                    | Bookkeeper Cluster  |
+| Bookie Write Rate              | meter_bookkeeper_bookie_write_rate                               | The bookie write rate (MB/min).                   | Bookkeeper Cluster  |
```
Comment (wankai123) [inline docs/en/setup/backend/backend-bookkeeper-monitoring.md:40]: The MAL rate means: `rate(Duration): Calculates the per-second average rate of increase in the time range.`
The `rate('PT1M')` does not mean per-min.

**Comment by wu-sheng** [comment]
@liangyepianzhou You can't change things blindly. It would only make you much slower.

File: test/e2e-v2/cases/pulsar/expected/metrics-has-value-instance-label.yml [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+        - key: _
+          value: *
+    values:
+      {{- contains .values }}
+      - id: {{ notEmpty .id }}
+        value: .
+        traceid: null
+      {{- end}}
+  {{- end}}
+error: null
```
Comment (wankai123) [inline test/e2e-v2/cases/pulsar/expected/metrics-has-value-instance-label.yml:30]: image
In fact, the e2e is failed. You used 1 expected file to verify all metrics and this expected file format is incorrect.
Please make sure no error in the e2e log.
you can refer the `vm's e2e`.

Comment (liangyepianzhou) [reply test/e2e-v2/cases/pulsar/expected/metrics-has-value-instance-label.yml:30]: This is due to the `value: *`.  Only the label part does not take effect.
```
type: TIME_SERIES_VALUES
results:
  {{- contains .results }}
  - metric:
      labels:
        - key: _
          value: *
    values:
      {{- contains .values }}
      - id: {{ notEmpty .id }}
        value: .
        traceid: null
      {{- end}}
  {{- end}}
error: null
```

Comment (liangyepianzhou) [reply test/e2e-v2/cases/pulsar/expected/metrics-has-value-instance-label.yml:30]: I will use two files to fix this.

Comment (liangyepianzhou) [reply test/e2e-v2/cases/pulsar/expected/metrics-has-value-instance-label.yml]: Fixed. There will be no error log.

**Review by wankai123 — APPROVED** [review]
LGTM
